torch/_subclasses/fake_tensor.py

```diff
@@ -1245,8 +1283,9 @@ def _cached_dispatch_impl(
             else:
                 self._validate_cache_key(func, args, kwargs)
                 output = self._dispatch_impl(func, types, args, kwargs)
-                entry = self._make_cache_entry(key, func, args, kwargs, output)
-                FakeTensorMode.cache[key] = entry
+                entry = self._make_cache_entry(state, key, func, args, kwargs, output)
+                key.strip_shape_env()
```

Comment:
would be great if we did not explicitly have to invoke `strip_shape_env` so that users of this could not get into a bad state. maybe either the key could do it in its `__init__` or this could be handled as in the dict as part of the `__setitem__` call

Reply:
I get the concern - but there's only one place where we need to call `strip_shape_env()` and that's just before we insert the value into the cache... so the danger area is pretty contained.

I hesitate to use `__setitem__` just because that adds a bunch of complexity/potential slowdown.